src/deployconfig.jl

```diff
@@ -843,6 +843,200 @@ authentication_method(::Buildkite) = Documenter.SSH
+        # Check if both are Ints. If both are Ints, then check if they are equal, otherwise, return false
+        pr_numbers_ok = all(x -> x isa Int, [pr_number1, pr_number2]) ? (pr_number1 == pr_number2) : false
+        is_pull_request_ok = get(ENV, "CI_BUILD_EVENT", "") == "pull_request"
+        pr_ok = pr_numbers_ok == is_pull_request_ok
+        all_ok &= pr_ok
+        println(io, "- $(marker(pr_numbers_ok)) ENV[\"CI_COMMIT_REF\"] corresponds to a PR")
+        println(io, "- $(marker(is_pull_request_ok)) ENV[\"CI_BUILD_EVENT\"] matches built type: `pull_request`")
+        btype_ok = push_preview
+        all_ok &= btype_ok
+        println(io, "- $(marker(btype_ok)) `push_preview` keyword argument to deploydocs is `true`")
+        deploy_branch = branch_previews
+        deploy_repo = repo_previews
+        is_preview = true
+        ## deploydocs to previews/PR
+        subfolder = "previews/PR$(something(pr_number1, 0))"
+    end
+
+    owner_ok = env_nonempty("CI_REPO_OWNER")
+    token_ok = env_nonempty("PROJECT_ACCESS_TOKEN")
+    auth_ok = owner_ok | token_ok
+    all_ok &= auth_ok
+
+    if token_ok
+        println(io, "- $(marker(token_ok)) ENV[\"PROJECT_ACCESS_TOKEN\"] exists and is non-empty")
+    end
+
+    if owner_ok
+        println(io, "- $(marker(owner_ok)) ENV[\"CI_REPO_OWNER\"] exists and is non-empty")
+    end
+
+    print(io, "Deploying: $(marker(all_ok))")
+    @info String(take!(io))
+    if build_type === :devbranch && !branch_ok && devbranch == "master" && cfg.woodpecker_ref == "refs/heads/main"
+        @warn """
+        Possible deploydocs() misconfiguration: main vs master. Current branch (from \$CI_COMMIT_REF) is "main". 
+        """
+    end
+
+    if all_ok
+        return DeployDecision(; all_ok=true,
+            branch=deploy_branch,
+            is_preview=is_preview,
+            repo=deploy_repo,
+            subfolder=subfolder
+        )
+    else
+        return DeployDecision(; all_ok=false)
+    end
+end
+
+authentication_method(::Woodpecker) = Documenter.HTTPS
+function authenticated_repo_url(cfg::Woodpecker)
+    if haskey(ENV, "FORGE_URL")
+        return "https://$(ENV["CI_REPO_OWNER"]):$(ENV["PROJECT_ACCESS_TOKEN"])@$(ENV["FORGE_URL"])/$(cfg.woodpecker_repo).git"
+    else
+        # Regex will return the root of the URL e.g. https://github.com/JuliaDocs/Documenter.jl → github.com
+        # `CI_REPO_LINK` is a built-in env var which returns a repository link e.g. https://github.com/JuliaDocs/Documenter.jl
+        # The variable is passed through the pipeline during the runs.
+        # See https://woodpecker-ci.org/docs/usage/environment to check the environment-variable 
+        m = match(r"https?:\/\/(?:.+\.)*(.+\..+?)\/", get(ENV, "CI_REPO_LINK", "")) 
+        forge_url = m !== nothing ? m.captures[1] : ""  # if `nothing`, just give an empty string so deploydocs can use it
+                                                        # even though we know that errors 
```

Comment:
An empty string is not really a sane value here though. This failure, should it occur, should be caught in `deploy_folder` and `all_ok` set to `false` there. I think you could add a `woodpecker_forge_url` to the struct, move this whole logic into the constructor, and check for the empty string in `deploy_folder`.